File: examples.ncs/getting-started/developing-with-ncs/1-cdb-py/packages/cdb/python/opercdbsub/opercdbsub.py
# -*- mode: python; python-indent: 4 -*-

import _ncs
import _ncs.cdb as cdb
import _ncs.maapi as maapi
from ncs_pyvm import NcsPyVM

import _namespaces.test_ns as ns

import socket
import threading

_schemas_loaded = False

# This Navu cdb subscriber subscribes to changes under the path
# /t:test/stats-item
# Whenever a change occurs there, the code iterates through the
# change and prints the values. The other examples dealing with
# configuration changes are triggered by the user adding/changing
# some configuration data. This cannot be done here, there is no
# way to manipulate operational data from the CLI. We must add
# that data using some other process.
# For CDB operational data, we can write a Java program which writes
# into the CDB operational store. The code in CreateOperData.java
# does that. Thus to invoke the main() in CreateOperData, we do
# # ant stats -Dop=CREATE -Dkey=k4
# in the directory above where the build.xml file resides.

# This will trigger the operational data subscriber below

# The code runs in an 'application' component, it implements
# the ApplicationComponent interface, this includes the run() method
# so the code will run in a Thread.


class OperSubscriber(object):
    def __init__(self, prio, path, debug):
        self.sock = socket.socket()
        self.path = path
        self.debug = debug
        self.prio = prio
        global _schemas_loaded

        if _schemas_loaded is False:
            ms = socket.socket()
            maapi.connect(
                sock=ms,
                ip='127.0.0.1',
                port=_ncs.NCS_PORT)
            maapi.load_schemas(ms)
            self.debug("Loaded schemas!\n")
            ms.close()
            _schemas_loaded = True

        cdb.connect(
            self.sock,
            type=cdb.DATA_SOCKET,
            ip='127.0.0.1',
            port=_ncs.NCS_PORT,
            path=self.path)
        self.subid = cdb.oper_subscribe(
            self.sock,
            path=self.path,
            nspace=ns.ns.hash)
        cdb.subscribe_done(self.sock)
        self.debug("Subscription {0}, subscribed to {1}".format(self.subid,
                                                                self.path))

    def wait(self):
        cdb.read_subscription_socket(self.sock)

    def ack(self):
        cdb.sync_subscription_socket(self.sock, cdb.DONE_OPERATIONAL)

    def diff_iter(self):
        def iterator(kp, op, oldv, newv, state):
            self.debug("kp : {0}, op : {1}, oldv : {2}, newv : {3}, state : {4}"
                       .format(str(kp),
                               op,
                               str(oldv),
                               str(newv),
                               str(state)))
            return _ncs.ITER_RECURSE
        cdb.diff_iterate(self.sock, self.subid, iterator, 0, None)

    def diff_iter_loop(self):
        while True:
            self.debug("waiting in iter loop...")
            self.wait()
            self.diff_iter()
            self.ack()

    def close(self):
        cdb.end_session(self.sock)
        cdb.close(self.sock)


# ---------------------------------------------
# COMPONENT THREAD THAT WILL BE STARTED BY NCS.
# ---------------------------------------------

class Opercdbsub(object):
    def __init__(self, *args, **kwds):
        # Setup the NCS object, containing mechanisms
        # for communicating between NCS and this User code.
        self._ncs = NcsPyVM(*args, **kwds)
        self.debug('--- Service INIT OBJECT')
        # Register our 'finish' callback
        self._finish_cb = lambda: self.finish()

        self.sub = OperSubscriber(prio=100, path="/test",
                                  debug=self.debug)

        self._ncs.reg_finish(self._finish_cb)

        self._stopevent = threading.Event()

    # This method is supposed to start the User application
    def run(self):
        self.debug("Runinng user application")
        self.sub.diff_iter_loop()

    # Just a convenient logging function
    def debug(self, line):
        self._ncs.debug(line)

    # Callback that will be invoked by NCS when the system is shutdown.
    # Make sure to shutdown the User code, including any User created threads.
    def finish(self):
        self.debug('OperSub in finish () =>\n')
        self.sub.close()
        self._stopevent.set()
        self.debug('OperSub in finish () => ok\n')

File: examples.ncs/getting-started/developing-with-ncs/1-cdb/packages/cdb/src/java/src/com/example/cdb/OperCdbSub.java
package com.example.cdb;

import java.util.Arrays;
import java.util.EnumSet;

import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.LogManager;

import com.example.cdb.namespaces.test;
import com.tailf.cdb.Cdb;
import com.tailf.cdb.CdbDBType;
import com.tailf.cdb.CdbDiffIterate;
import com.tailf.cdb.CdbSession;
import com.tailf.cdb.CdbSubscription;
import com.tailf.cdb.CdbSubscriptionSyncType;
import com.tailf.cdb.CdbSubscriptionType;
import com.tailf.conf.ConfObject;
import com.tailf.conf.DiffIterateFlags;
import com.tailf.conf.DiffIterateOperFlag;
import com.tailf.conf.DiffIterateResultFlag;
import com.tailf.ncs.ApplicationComponent;
import com.tailf.ncs.ResourceManager;
import com.tailf.ncs.annotations.Resource;
import com.tailf.ncs.annotations.ResourceType;
import com.tailf.ncs.annotations.Scope;

// This Navu cdb subscriber subscribes to changes under the path
// /t:test/stats-item
// Whenever a change occurs there, the code iterates through the
// change and prints the values. The other examples dealing with
// configuration changes are triggered by the user adding/changing
// some configuration data. This cannot be done here, there is no
// way to manipulate operational data from the CLI. We must add
// that data using some other process.
// For CDB operational data, we can write a Java program which writes
// into the CDB operational store. The code in CreateOperData.java
// does that. Thus to invoke the main() in CreateOperData, we do
// # ant stats -Dop=CREATE -Dkey=k4
// in the directory above where the build.xml file resides.

// This will trigger the operational data subscriber below

// The code runs in an 'application' component, it implements
// the ApplicationComponent interface, this includes the run() method
// so the code will run in a Thread.

public class OperCdbSub implements ApplicationComponent, CdbDiffIterate  {
    private static Logger LOGGER = LogManager.getLogger(OperCdbSub.class);

    // let our ResourceManager inject Cdb sockets to us
    // no explicit creation of creating and opening sockets needed
    @Resource(type=ResourceType.CDB, scope=Scope.INSTANCE, qualifier="sub-sock")
    private Cdb cdbSub;
    @Resource(type=ResourceType.CDB,scope=Scope.INSTANCE, qualifier="data-sock")
    private Cdb cdbData;

    private boolean requestStop = false;
    private int point;
    private CdbSubscription cdbSubscription;

    public OperCdbSub() {
    }

    public void init() {
        LOGGER.info(" init oper subscriber ");
        try {
            cdbSubscription = cdbSub.newSubscription();
            String path = "/t:test/stats-item";
            point =
                cdbSubscription.subscribe(CdbSubscriptionType.SUB_OPERATIONAL,
                                          1, test.hash, path);
            cdbSubscription.subscribeDone();
            LOGGER.info("subscribeDone");
            requestStop = false;
        } catch (Exception e) {
            LOGGER.error("Fail in init",e);
        }
    }

    public void run() {
        try {
            while (!requestStop) {
                try {
                    int[] point = cdbSubscription.read();
                    CdbSession cdbSession =
                        cdbData.startSession(CdbDBType.CDB_OPERATIONAL);
                    EnumSet<DiffIterateFlags> diffFlags =
                        EnumSet.of(DiffIterateFlags.ITER_WANT_PREV);
                    cdbSubscription.diffIterate(point[0], this, diffFlags,
                                                cdbSession);
                    cdbSession.endSession();
                } finally {
                    cdbSubscription.sync(
                                    CdbSubscriptionSyncType.DONE_OPERATIONAL);
                }
            }
        } catch (Exception e) {
            LOGGER.error("Fail in run shouldrun", e);
        }
        requestStop = false;
    }

    public void finish() {
        requestStop = true;
        try {
            ResourceManager.unregisterResources(this);
        } catch (Exception e) {
            LOGGER.error("Fail in finish", e);
        }
    }

    @Override
    public DiffIterateResultFlag iterate(ConfObject[] kp,
                                         DiffIterateOperFlag op,
                                         ConfObject oldValue,
                                         ConfObject newValue,
                                         Object initstate) {
        LOGGER.info(op + " " + Arrays.toString(kp) +" value: " + newValue);
        switch (op){
            case MOP_DELETED:
                break;
            case MOP_CREATED:
            case MOP_MODIFIED: {
                break;
            }
        }
        return DiffIterateResultFlag.ITER_RECURSE;
    }
}
